File: API/Utility/PasswordPolicy.cs
using API.Exceptions;
using System.Text.RegularExpressions;

namespace API.Utility
{
    /// <summary>Reglas de contraseña para cambio de credenciales (alineadas con políticas fuertes habituales).</summary>
    public static class PasswordPolicy
    {
        public const int MinLength = 12;
        public const int MaxLength = 128;

        /// <summary>Carácter especial ASCII típico en políticas empresariales.</summary>
        private static readonly Regex HasSpecial = new(@"[!@#$%^&*()_+\-=\[\]{};':""\\|,.<>\/?~`]", RegexOptions.Compiled);

        private static readonly Regex HasUpper = new(@"[A-Z]", RegexOptions.Compiled);
        private static readonly Regex HasLower = new(@"[a-z]", RegexOptions.Compiled);
        private static readonly Regex HasDigit = new(@"\d", RegexOptions.Compiled);

        public static void ValidateNewPassword(string? newPassword)
        {
            if (string.IsNullOrWhiteSpace(newPassword))
                throw new BadRequestException("La nueva contraseña es obligatoria.");

            if (newPassword.Length < MinLength)
                throw new BadRequestException($"La contraseña debe tener al menos {MinLength} caracteres.");

            if (newPassword.Length > MaxLength)
                throw new BadRequestException($"La contraseña no puede superar {MaxLength} caracteres.");

            if (!HasUpper.IsMatch(newPassword))
                throw new BadRequestException("La contraseña debe incluir al menos una letra mayúscula (A-Z).");

            if (!HasLower.IsMatch(newPassword))
                throw new BadRequestException("La contraseña debe incluir al menos una letra minúscula (a-z).");

            if (!HasDigit.IsMatch(newPassword))
                throw new BadRequestException("La contraseña debe incluir al menos un dígito (0-9).");

            if (!HasSpecial.IsMatch(newPassword))
                throw new BadRequestException(
                    "La contraseña debe incluir al menos un símbolo (por ejemplo: ! @ # $ % ^ & * _ - = + [ ] { }).");
        }
    }
}

File: UI/src/app/utils/password-policy.ts
/** Misma política que API/Utility/PasswordPolicy.cs */
export const PASSWORD_MIN_LENGTH = 12;
export const PASSWORD_MAX_LENGTH = 128;

const hasUpper = /[A-Z]/;
const hasLower = /[a-z]/;
const hasDigit = /\d/;
const hasSpecial = /[!@#$%^&*()_+\-=[\]{};':"\\|,.<>/?~`]/;

/** Si no es válida, devuelve el mensaje de error; si es válida, null. */
export function validateStrongPassword(password: string): string | null {
  const p = password ?? '';
  if (!p.trim()) {
    return 'La nueva contraseña es obligatoria.';
  }
  if (p.length < PASSWORD_MIN_LENGTH) {
    return `La contraseña debe tener al menos ${PASSWORD_MIN_LENGTH} caracteres.`;
  }
  if (p.length > PASSWORD_MAX_LENGTH) {
    return `La contraseña no puede superar ${PASSWORD_MAX_LENGTH} caracteres.`;
  }
  if (!hasUpper.test(p)) {
    return 'La contraseña debe incluir al menos una letra mayúscula (A-Z).';
  }
  if (!hasLower.test(p)) {
    return 'La contraseña debe incluir al menos una letra minúscula (a-z).';
  }
  if (!hasDigit.test(p)) {
    return 'La contraseña debe incluir al menos un dígito (0-9).';
  }
  if (!hasSpecial.test(p)) {
    return 'La contraseña debe incluir al menos un símbolo (por ejemplo: ! @ # $ % ^ & * _ - = + [ ] { }).';
  }
  return null;
}
